### DISCUSSION TITLE
How to combine User::where and User::select

### DISCUSSION THREAD

**Discussion by shark2k** [DirectoryTree/LdapRecord#565] [Q&A] [answered]
Hello,

I'm hoping that this is possible to do, I just have no idea how I would go about accomplishing it.

I'm writing some code to test this and right now what I have is the following (again just for testing purposes):

```php
$employees = User::where('c', '=', 'US')->paginate();
foreach ($employees as $employee) {
     echo $employee->getFirstAttribute('sn').' '.$employee->getFirstAttribute('givenName').' - '.$employee->getFirstAttribute('c');
     echo '<br />------------------------------------------------------<br />';
}
```

Based on another answer found in the issues section I believe, I found that you can do ```User::select(['attribute1', 'attribute2'])->get()``` (or paginate or even into a chunk) where attribute1 and attribute2 would be the actual name adding all the attributes you want to reduce the amount of data being retrieved.

I like the select feature specifically because it reduces the amount of data you are getting, but I'm having trouble figuring out how I can combine the ```User::where``` and the ```User::select``` features together.

Leaving the below just for the history, but I realized the issue was my fault in my one if statement where I was checking for the givenName to be equal to blank, so of course I wasn't outputting anything.  The above I am still unsure of though.

<strike>Also, I tried combining the ```User::where``` with my chunking code that worked with the ```User::select``` but I did not get any results.

The way that code looked like was the following (since I'm guessing I did something wrong):

```php
User::where('c', '=', 'US')->chunk(1000, function ($employees) {
     foreach ($employees as $employee) {
          $username = $employee->getFirstAttribute('samaccountname');

          if ($employee->getFirstAttribute('userAccountControl') == 512) {
               if ($employee->getFirstAttribute('givenName') == '') {
                    echo $username.','.$employee->getFirstAttribute('givenName').','.$employee->getFirstAttribute('displayName').','.$employee->getFirstAttribute('countryCode').','.$employee->getFirstAttribute('whenChanged').'<br />';
               }
          }
     }
});
```

If I change the ```User::where``` to a ```User::select(['samaccountname','givenName','displayName','countryCode','whenChanged','userAccountControl','pfUniqueID'])'``` in the prior code, I get results.

After getting the ```User::where``` to work by using it with ```paginate()``` I'm thinking it has something to do with the fact that I am not assigning it to a variable, unless that is an automatic thing with the chunking (I'm not really clear on that)?</strike>

Appreciate any direction anyone can give me on how to actually accomplish this (again if it's possible).

Thanks,
Shark2k

**Comment by shark2k** [comment, marked as answer]
Alright, so answering this myself as I ended up trying something (which I should've tried before opening this discussion, so my bad).

The answer was as simple as doing one of the following:

```php
$employees = User::where('c', '=', 'US')->select(['samaccountname','sn','givenName'])->get();
```
or
```php
$employees = User::select(['samaccountname','sn','givenName'])->where('c', '=', 'US')->get();
```

Not being super comfortable with OOP, I wasn't sure if they could be linked together (and for some reason initially had a some reservations about how exactly to do the "linking" because I couldn't get past how they were being called, though looking at get it should have been clear).

Anyway, decided to try both ways above and they both got me the results I wanted.
Now, the only thing I am curious about is whether one way is better then the other.
Part of me is thinking the first way is better because it's filtering down the overall list to only specific entries with the 'c' value and then of those it's only taking the attributes that I want.
However, that's more of a curiosity then needing to actually be answered.

-Shark2k

**Reply by stevebauman** [reply]
Hi @shark2k, apologies for the delay with a reply.

Your solutions are correct. The query builder is designed in a fluent way. This means, method chaining is intended, and it is a feature of the builder.

Both of your posted solutions will return identical results, it's up to you which you prefer to be more readable and applicable to your application.

If you're using an IDE that supports type-hinting, it may help understand which methods can be called by first calling the static method `query()` on the `User` model. Your IDE will show you methods that can be chained, and display the appropriate return type and supported parameters:

```php
use LdapRecord\Models\ActiveDirectory\User;

User::query()->select()->where('...')->get();
```

Calling `get()` at the end of the built query compiles all of the chained data that has been set on the query builder, then sends the request to your LDAP server. Order of the methods that are called are not important, unless you call them multiple times.

For example, calling `select()` will overwrite the previous `select()`:

```php
// Will result in only "bar" being selected, as it will overwrite the previous select call.
User::select('foo')->select('bar');
```

However, this is not the case with `where()` clauses:

```php
// Search for a user with the first name of "John" and last name of "Doe":
User::where('givenname', '=', 'John')->where('surname', '=', 'Doe')->get();
```

I hope this helps! Let me know if you have any further questions.

**Reply by shark2k** [reply]
@stevebauman No worries on the delay with the reply.

I actually figured it would be doable and really should have tried something before posting the question.
Since I had posted the question though, I figured I'd answer myself just in case someone else was uncertain in the future and was looking for an answer.

That said, your additional answer is actually super helpful as well as you explained it in more detail (and also answered my question/concern within my answer about order).

Thanks for taking the time to reply with the additional information.

-Shark2k

**Reply by stevebauman** [reply]
Happy to help @shark2k! I'm glad you were able to resolve it and I appreciate you posting the solution for others ❤️
